com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/eval/JdtEvaluationProvider.java

```diff
@@ -0,0 +1,273 @@
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.debug.eval.ICompiledExpression;
+import org.eclipse.jdt.internal.debug.core.model.JDIDebugTarget;
+import org.eclipse.jdt.internal.debug.core.model.JDIStackFrame;
+import org.eclipse.jdt.internal.debug.core.model.JDIThread;
+import org.eclipse.jdt.internal.debug.eval.ast.engine.ASTEvaluationEngine;
+import org.eclipse.jdt.internal.launching.JavaSourceLookupDirector;
+
+import com.microsoft.java.debug.core.Configuration;
+import com.microsoft.java.debug.core.adapter.IEvaluationProvider;
+import com.sun.jdi.StackFrame;
+import com.sun.jdi.ThreadReference;
+import com.sun.jdi.Value;
+
+public class JdtEvaluationProvider implements IEvaluationProvider {
+    private static final Logger logger = Logger.getLogger(Configuration.LOGGER_NAME);
+    private IJavaProject project;
+    private ILaunch launch;
+    private JDIDebugTarget debugTarget;
+    private Map<ThreadReference, JDIThread> threadMap = new HashMap<>();
+
+    @Override
+    public CompletableFuture<Value> evaluate(String projectName, String code, StackFrame sf) {
+        CompletableFuture<Value> completableFuture = new CompletableFuture<>();
+        if (debugTarget == null) {
+            if (project == null) {
+                if (StringUtils.isBlank(projectName)) {
+                    // TODO: get project from stackframe
+                    logger.severe("Cannot evaluate when project is not specified.");
+                    completableFuture
+                    .completeExceptionally(new IllegalStateException("Please specify projectName in launch.json."));
+                } else {
+                    for (IProject proj : ResourcesPlugin.getWorkspace().getRoot().getProjects()) {
+                        try {
+                            if (proj.getName().equals(projectName)) {
+                                if (!proj.isNatureEnabled("org.eclipse.jdt.core.javanature")) {
+                                    completableFuture.completeExceptionally(
+                                            new IllegalStateException(String.format("Project %s is not a java project.", projectName)));
+                                    return completableFuture;
+                                }
+                                project = JavaCore.create(proj);
+                                break;
+                            }
+                        } catch (CoreException e) {
+                            logger.severe(String.format("Cannot initialize project: %s", e.toString()));
+                        }
+                    }
+                }
+            }
+
+            if (project == null) {
+                completableFuture.completeExceptionally(new IllegalStateException(String.format("Project %s cannot be found.", projectName)));
+                return completableFuture;
+            }
+            if (launch == null) {
```

Comment:
If launch != null, then we have an issue here. That means we already have a ILaunchMock created before and we are reusing the same provider.

Reply:
provider is not shared cross debug session, it will serve for only one debug session in current impl.